### PR TITLE
fix(moe): add the replicated shared expert outside the TP all_reduce

### PR DESCRIPTION
**Pull request by rebel-kblee**
### 🚀 Summary of Changes

fix to support tp > 1, dp > 1 in deepseek

> *What does this PR do? What feature, fix, or improvement does it bring?*

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [x] 🛠 Bug fix (`fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [x] PR title follows Conventional Commits format
* [x] This PR is linked to an existing issue
* [x] The test method is described, and the expected result is clearly stated
* [x] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

File: vllm_rbln/patches/deepseek_v2.py [hunk 1]
```diff
@@ -38,14 +38,20 @@ def patched_deepseek_v2_moe_forward(
...
-        final_hidden_states = final_hidden_states + shared_output
+        if not self.is_sequence_parallel:
+            final_hidden_states = final_hidden_states + shared_output
+            shared_output = None
 
     if self.tp_size > 1:
         final_hidden_states = tensor_model_parallel_all_reduce(final_hidden_states)
+
+    if shared_output is not None:
+        final_hidden_states = final_hidden_states + shared_output
```
Comment (rebel-cjlee) [inline vllm_rbln/patches/deepseek_v2.py:54]: Just curious, I think we don't typically use sequence parallelism, so it looks the same as before. Did we have any problems with our previous TP setup?

**Review by rebel-cjlee — APPROVED** [review]

File: vllm_rbln/patches/deepseek_v2.py (hunk 1 above)
Comment (rebel-kblee) [reply vllm_rbln/patches/deepseek_v2.py:54]: When TP>1, DP>1 with EP -> backend set to True
(https://github.com/vllm-project/vllm/blob/ee0da84ab9e04ac7610e28580af62c365e898389/vllm/config/parallel.py#L642-L656)
Also, deepseek v2 model set this config in shared_experts
(https://github.com/vllm-project/vllm/blob/ee0da84ab9e04ac7610e28580af62c365e898389/vllm/model_executor/models/deepseek_v2.py#L266)
Therefore, as we ultimately need to distinguish between the cases `TP`, `DP` and `TP+DP `in order to decide whether to use `all_reduce` or `all_gather`, this adjustment is necessary.

Comment (rebel-kblee) [reply vllm_rbln/patches/deepseek_v2.py:54]: @rebel-cjlee Would it be all right if I asked you to double-check this?

Comment (rebel-cjlee) [reply vllm_rbln/patches/deepseek_v2.py:54]: Checked this against v0.24.0 and you're right, the fix is needed.
